inferno-ml-server-types/src/Inferno/ML/Server/Types.hs

```diff
@@ -0,0 +1,762 @@
+    where
+      thumbnailP :: Maybe Object -> Parser (Maybe (Text, URIRef Absolute))
+      thumbnailP = \case
+        Nothing -> pure Nothing
+        Just o -> fmap Just $ (,) <$> o .: "description" <*> o .: "url"
+
+instance ToJSON ModelMetadata where
+  toJSON =
+    genericToJSON
+      defaultOptions
+        { fieldLabelModifier = camelTo2 '_',
+          omitNothingFields = True
+        }
+
+-- | Similar to the @Version@ type from base, but allows for a leading @v@ and
+-- guarantees that there is at least one digit. Digits must be separated by @.@;
+-- multiple tags are allowed, separated by @-@
+data Version
+  = Version
+      (NonEmpty Int)
+      -- ^ List of digits for version string
+      [Text]
+      -- ^ Any tags
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NFData)
+
+-- Compares based on digits, not on tag
+instance Ord Version where
+  compare = comparing . view $ typed @(NonEmpty Int)
+
+instance FromJSON Version where
+  parseJSON =
+    withText "Version" $
+      either fail pure
+        . Attoparsec.parseOnly versionP
+        . Text.Encoding.encodeUtf8
+
+instance ToJSON Version where
+  toJSON = String . showVersion
+
+instance FromField Version where
+  fromField f =
+    maybe (returnError UnexpectedNull f mempty) $
+      either (returnError ConversionFailed f) pure
+        . Attoparsec.parseOnly versionP
+
+instance ToField Version where
+  toField = Escape . Text.Encoding.encodeUtf8 . showVersion
+
+versionP :: Attoparsec.Parser Version
+versionP = do
+  void . optional $ Attoparsec.char 'v'
+  Version <$> numbersP <*> tagsP
+  where
+    numbersP :: Attoparsec.Parser (NonEmpty Int)
+    numbersP =
+      -- Note that `sepBy1` will fail if there is not at least one occurrence,
+      -- so using `fromList` is OK here
+      fmap NonEmpty.fromList
+        -- Note that if `many1` succeeds, `read` must also succeed
+        . Attoparsec.sepBy1 (read <$> Attoparsec.many1 Attoparsec.digit)
+        $ Attoparsec.char '.'
+
+    tagsP :: Attoparsec.Parser [Text]
+    tagsP =
+      fmap (filter (not . Text.null) . fmap Text.Encoding.decodeUtf8)
+        . Attoparsec.sepBy
+          ( Attoparsec.takeWhile (Attoparsec.inClass "a-zA-Z")
+          )
+        $ Attoparsec.char '-'
+
+showVersion :: Version -> Text
+showVersion (Version ns ts) =
+  mconcat
+    [ "v",
+      Text.intercalate "." . fmap tshow $ NonEmpty.toList ns,
+      bool ("-" <> Text.intercalate "-" ts) mempty $ null ts
+    ]
+
+-- | Row of the inference parameter table, parameterized by the user, group, and
+-- script type
+data InferenceParam uid gid p s = InferenceParam
+  { id :: Maybe (Id (InferenceParam uid gid p s)),
+    -- | The script of the parameter
+    --
+    -- For new parameters, this will be textual or some other identifier
+    -- (e.g. a UUID for use with @inferno-lsp@)
+    --
+    -- For existing inference params, this is the foreign key for the specific
+    -- script in the 'InferenceScript' table
```

Comment:
Curious about this -- it looks like it's for backwards compatibility but we haven't yet launched inferno-ml -- is this something we intend to remove before the public launch?

Reply:
Actually, it's so that we can support posting all of the information to create an inference param without needing to have a `script` field be a `VCObjectHash`. So for example, it could be the UUID of a script parsed by `inferno-lsp` 

By "existing" param, I mean when we save a param to the DB, this field will be a `VCObjectHash`, not that there are params saved somewhere already that we need to maintain backwards compatibility with. I will rephrase it to make that clearer